hudi-spark/src/main/java/org/apache/hudi/keygen/CustomKeyGenerator.java

```diff
@@ -0,0 +1,128 @@
+import org.apache.hudi.exception.HoodieDeltaStreamerException;
+import org.apache.hudi.exception.HoodieKeyException;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * This is a generic implementation of KeyGenerator where users can configure record key as a single field or a combination of fields.
+ * Similarly partition path can be configured to have multiple fields or only one field. This class expects value for prop
+ * "hoodie.datasource.write.partitionpath.field" in a specific format. For example:
+ *
+ * properties.put("hoodie.datasource.write.partitionpath.field", "field1:PartitionKeyType1,field2:PartitionKeyType2").
+ *
+ * The complete partition path is created as <value for field1 basis PartitionKeyType1>/<value for field2 basis PartitionKeyType2> and so on.
+ *
+ * Few points to consider:
+ * 1. If you want to customise some partition path field on a timestamp basis, you can use field1:timestampBased
+ * 2. If you simply want to have the value of your configured field in the partition path, use field1:simple
+ * 3. If you want your table to be non partitioned, simply leave it as blank.
+ *
+ * RecordKey is internally generated using either SimpleKeyGenerator or ComplexKeyGenerator.
+ */
+public class CustomKeyGenerator extends KeyGenerator {
+
+  protected final List<String> recordKeyFields;
+  protected final List<String> partitionPathFields;
+  protected final TypedProperties properties;
+  private static final String DEFAULT_PARTITION_PATH_SEPARATOR = "/";
+  private static final String SPLIT_REGEX = ":";
+
+  /**
+   * Used as a part of config in CustomKeyGenerator.java.
+   */
+  public enum PartitionKeyType {
+    SIMPLE, TIMESTAMP
+  }
+
+  public CustomKeyGenerator(TypedProperties props) {
+    super(props);
+    this.properties = props;
+    this.recordKeyFields = Arrays.stream(props.getString(DataSourceWriteOptions.RECORDKEY_FIELD_OPT_KEY()).split(",")).map(String::trim).collect(Collectors.toList());
+    this.partitionPathFields =
+      Arrays.stream(props.getString(DataSourceWriteOptions.PARTITIONPATH_FIELD_OPT_KEY()).split(",")).map(String::trim).collect(Collectors.toList());
+  }
+
+  @Override
+  public HoodieKey getKey(GenericRecord record) {
+    //call function to get the record key
+    String recordKey = getRecordKey(record);
+    //call function to get the partition key based on the type for that partition path field
+    String partitionPath = getPartitionPath(record);
+    return new HoodieKey(recordKey, partitionPath);
+  }
+
+  public String getPartitionPath(GenericRecord record) {
+    if (partitionPathFields == null) {
+      throw new HoodieKeyException("Unable to find field names for partition path in cfg");
+    }
+
+    String partitionPathField;
+    StringBuilder partitionPath = new StringBuilder();
+
+    //Corresponds to no partition case
+    if (partitionPathFields.size() == 1 && partitionPathFields.get(0).isEmpty()) {
```

Comment:
sorry, if size = 1, how is it a no partition case ?

Reply:
So the idea behind this was if user wants to not have any partitions, he can simply leave partitionPathFields as blank. If he leaves the property value blank, that corresponds to a list with one element and that element is empty string. You can find more context here - https://github.com/apache/incubator-hudi/pull/1433#discussion_r414306395

Reply:
I get it, but have a follow up question and a comment. 
- I am not sure why would someone set empty string. if users wants to not have any partitions, might as well not set it only ? Anyways, guess you discussed w/ vinoth on that. Will let you folks decide on whats better. 
- Assuming, user might set it to empty, in line 69 where we generate  partitionPathFields, we trim for empty strings isn't? So, the list should be empty in that case in my understanding. correct me if I am wrong.

Reply:
> if users wants to not have any partitions, might as well not set it only

Actually in the constructor of CustomKeyGenerator, partitionPathFields get initialised through a call to props.getString() which internally checks if key exists. So user will have to include this key always. I am ok with both ways, so chose to follow what Vinoth suggested. 

> So, the list should be empty in that case in my understanding. correct me if I am wrong

No the list actually contains one element always which is an empty string. I have verified this. :)

Reply:
ok, my bad. Just checked the trim docs, it will just trim leading and trailing whitespaces if any.